### PR TITLE
Implement none-behavior strategies

### PR DESCRIPTION
**Pull request by grst**
This PR adds a new option `none_behavior` that allows to specify different strategies for handling `None` values. 
For now there are only two strategies implemented 
 * `NONE_BEHAVIOR_DEFAULT` -> attempt to merge or fail (previous hyapyco behavior)
 * `NONE_BEHAVIOR_OVERRIDE` -> `None` overrides the existing value, independent of the merge strategy. 

Other strategies could be implemented in the future such as ignoring None values.

### CODE REVIEW

File: hiyapyco/__init__.py [hunk 1]
```diff
@@ -119,6 +124,19 @@ def __init__(self, *args, **kwargs):
...
+            logger.debug("parse kwarg method: %s ..." % kwargs["none_behavior"])
+            if kwargs["none_behavior"] not in NONE_BEHAVIORS.values():
+                raise HiYaPyCoInvocationException(
+                    "undefined method used, must be one of: %s"
+                    % " ".join(NONE_BEHAVIORS.keys())
+                )
+            self.none_behavior = kwargs["none_behavior"]
+            del kwargs["none_behavior"]
+        if self.none_behavior is None:
+            self.none_behavior = NONE_BEHAVIOR_DEFAULT
```
Comment (grst) [inline hiyapyco/__init__.py:138]: Is there a reason why arguments are evaulated in such a (IMO) cumbersome way? This could be just a keyword argument to the function instead of manually fiddling with `kwargs`... I would probably have implemented it like that when starting from scratch:

```python
NoneBehavior = Enum("NoneBehavior", ["DEFAULT", "OVERRIDE"])

def __init__(self, *args, none_behavior: NoneBehavior = None, **kwargs):
    if none_behavior not in NoneBehavior:
        raise HiYaPyCoInvocationException(f"Undefined non_behavior, must be one of {', '.join(NoneBehavior._member_names_}")
    self.none_behavior = none_behavior
```

**Comment by zerwes** [comment]
I am currently on vacation and will take care of it when I get back ...

File: hiyapyco/__init__.py (hunk 1 above)
Comment (zerwes) [reply hiyapyco/__init__.py:138]: @grst ... I see what you mean, but to be honest I can not remember what idea or cerebral nod motivated me to implement it in such a detailed and expansive notation approx 10 years ago ... to be honest, I had no idea about the widespread use and longevity of the project during the first implementation that was more intended for my private use
Anyhow, just to not let the topic drown, I opened #81, thank you for your hint.

Comment (grst) [reply hiyapyco/__init__.py:138]: 10 years is also a long time ago and many of the newer language features like type hints or enums didn't exist. But nowadays Python 3.9 is the [oldest supported version](https://devguide.python.org/versions/) :)

File: hiyapyco/__init__.py [hunk 2]
```diff
@@ -327,11 +345,23 @@ def _simplemerge(self, a, b):
...
+        logger.debug(
+            "simplemerge %s (%s) and %s (%s)"
+            % (
+                a,
+                type(a),
+                b,
+                type(b),
+            )
+        )
         if b is None:
```
Comment (zerwes) [inline hiyapyco/__init__.py:357]: would you mind changing it to something like
```diff
-           logger.debug('pass as b is None')
            # override -> None replaces object, no matter what.
            if self.none_behavior == NONE_BEHAVIOR_OVERRIDE:
+               logger.debug('b is None + none_behavior in use => return None')
                return None
            # default behavior is to attempt merge or fail
-            else:
+            logger.debug('pass as b is None')
-               pass
+            pass
```
avoids a unnecessary `else` and improves debug output ...

File: hiyapyco/__init__.py [hunk 3]
```diff
@@ -366,11 +396,21 @@ def _substmerge(self, a, b):
             a = copy.deepcopy(a)
             b = copy.deepcopy(b)
         logger.debug('>' * 30)
-        logger.debug('substmerge %s and %s' % (a, b,))
-        # FIXME: make None usage configurable
+        logger.debug(
```
Comment (zerwes) [inline hiyapyco/__init__.py:399]: I do not think that this change is related to the topic of the PR and it does not improve readability ... pls revert this change

File: hiyapyco/__init__.py [hunk 4]
```diff
@@ -327,11 +345,23 @@ def _simplemerge(self, a, b):
...
+                a,
+                type(a),
+                b,
+                type(b),
+            )
+        )
         if b is None:
             logger.debug('pass as b is None')
-            pass
+            # override -> None replaces object, no matter what.
```
Comment (zerwes) [inline hiyapyco/__init__.py:359]: same as in the previous comment

File: hiyapyco/__init__.py [hunk 5]
```diff
@@ -366,11 +396,21 @@ def _substmerge(self, a, b):
...
+            "substmerge %s and %s"
+            % (
+                a,
+                b,
+            )
+        )
         if b is None:
             logger.debug('pass as b is None')
-            pass
+            # override -> None replaces object, no matter what.
```
Comment (zerwes) [inline hiyapyco/__init__.py:408]: same as in the first comment of the review

File: hiyapyco/__init__.py [hunk 6]
```diff
@@ -421,11 +461,21 @@ def _deepmerge(self, a, b, context = None):
...
+            "deepmerge %s and %s"
+            % (
+                a,
+                b,
+            )
+        )
         if b is None:
             logger.debug('pass as b is None')
-            pass
+            # override -> None replaces object, no matter what.
```
Comment (zerwes) [inline hiyapyco/__init__.py:473]: same as in the first comment of the review

File: hiyapyco/__init__.py [hunk 7]
```diff
@@ -421,11 +461,21 @@ def _deepmerge(self, a, b, context = None):
         if self.dereferenceyamlanchors:
             a = copy.deepcopy(a)
             b = copy.deepcopy(b)
-        logger.debug('deepmerge %s and %s' % (a, b,))
-        # FIXME: make None usage configurable
+        logger.debug(
```
Comment (zerwes) [inline hiyapyco/__init__.py:464]: I do not think that this change is related to the topic of the PR and it does not improve readability ... pls revert this change

**Review by zerwes — CHANGES_REQUESTED** [review]
Hello @grst 
Thank you for the PR taking care of my `FIXME: make None usage configurable` notes! I never found the time (and had no urgent need to) fix this.
I just have proposed some cosmetic changes, but the functionality itself LGTM. Thank you.

File: hiyapyco/__init__.py (hunk 2 above)
Comment (grst) [reply hiyapyco/__init__.py:349]: Makes sense! 

btw, when using 

````
```suggestion
```
````

image


instead of 
````
```diff
```
````

one can directly apply the changes with one click. 
Because you were wondering, [here](https://github.com/zerwes/hiyapyco/pull/76#issuecomment-2317209443) ;)

File: hiyapyco/__init__.py (hunk 3 above)
Comment (grst) [reply hiyapyco/__init__.py]: sorry, this was caused by the ruff autoformatter which I forgot to turn off. I like autoformatting a lot, especially because it prevents discussions like this during PR review, but I agree this change doesn't make sense in isolation.

File: hiyapyco/__init__.py (hunk 2 above)
Comment (zerwes) [reply hiyapyco/__init__.py:354]: Thank you for the hint ... looks like not only my python coding style, but my github knowledge too is more then slightly outdated :-1: (probably much more stuff including my self, but this is another story ...)

**Review by zerwes — CHANGES_REQUESTED** [review]
sorry @grst 
I assume caused by the usage of _diff_ instead of _suggest_ the original `pass` statement has been lost ...

File: README.rst [hunk 8]
```diff
@@ -139,6 +139,11 @@ kwargs
 
 - ``mergelists``: boolean try to merge lists of dict (default: ``True``)
 
+- ``none_behavior``: bit (one of the listed below):
+
+   -  ``hiyapyco.NONE_BEHAVIOR_DEFAULT``: attempt to merge the value with ``None`` and fail if this is not possible (default method)
+   -  ``hiyapyco.NONE_BEHAVIOR_OVERRIDE``: ``None`` always overrides any other value.
```
Comment (zerwes) [inline README.rst:145]:
```suggestion
   -  ``hiyapyco.NONE_BEHAVIOR_OVERRIDE``: ``None`` always overrides any other value.
```

File: test/base_none_behavior.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4 smartindent
+singel: null
```
Comment (zerwes) [inline test/base_none_behavior.yaml:2]:
```suggestion
singel: null
```

File: test/base_none_behavior.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4 smartindent
+singel: null 
+int: null
```
Comment (zerwes) [inline test/base_none_behavior.yaml:3]:
```suggestion
int: null
```

File: test/base_none_behavior.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4 smartindent
+singel: null 
+int: null 
+array: null
+hash: null
+deeplist:
+    - d1: null
+    - d2:
+        d2k1: null
```
Comment (zerwes) [inline test/base_none_behavior.yaml:9]:
```suggestion
        d2k1: null
```

File: test/base_none_behavior.yaml [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+array: null
+hash: null
+deeplist:
+    - d1: null
+    - d2:
+        d2k1: null 
+        d2k2: x2
+deepmap:
+    l1k1:
+        l2k1: null
```
Comment (zerwes) [inline test/base_none_behavior.yaml:13]:
```suggestion
        l2k1: null
```

**Review by zerwes — CHANGES_REQUESTED** [review]
sorry, the cli diff revealed some white spaces ...

**Comment by grst** [comment]
good catch, seems I got too much used to autoformatting 😅

**Review by zerwes — APPROVED** [review]
